src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res.refs = repeat(res.refs, inner = [v.inner], outer = [v.outer])
     res
 end
+
+
+"""
+    transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+
+Transpose a `DataFrame`, such that rows become columns,
+and the column indexed by `src_namescol` becomes a header.
+
+By default, the type of resulting columns will depend on the promoted type of all input columns.
+Pass `promote_type=false` to make resulting column types depend on the elements of the originating rows,
+though note that this may be substantially slower.
+
+
+# Examples
+
+```julia
+df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2)) # all types can be promoted to Float64
+df2 = DataFrame(a=["x", "y"], b=[1, "str"], c=[1,2], d=rand(Bool,2))
+
+transpose(df1)
+transpose(df1, promote_type=false)
+
+transpose(df2)
+transpose(df2, promote_type=false)
+````
+"""
+function Base.transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
```

Comment:
having `src_namescol` and `dest_namescol` positional is OK I think. I am not 100% sure about the defaults (and if we should provide them, especially for `dest_namescol`), but let us keep what you propose. `copycols` is not needed. Here is a proposal of a signature:
```suggestion
function Base.transpose(df::AbstractDataFrame, src_namescol::ColumnIndex=1,
                        dest_namescol::Union{Symbol, AbstractString}=src_namescol isa Integer ?
                                                                     _names(df)[src_namescol] :
                                                                     src_namescol;
                        makeunique::Bool=false, promote_type::Bool=true)
```